File: WasteLessServer/src/main/java/server/mediator/Mediator.java
package server.mediator;

import org.jetbrains.annotations.NotNull;
import org.springframework.beans.BeansException;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.stereotype.Component;
import server.mediator.command.*;
import server.mediator.handler.*;
import server.mediator.query.*;

import java.util.HashMap;
import java.util.Map;

@Component
public class Mediator implements ApplicationContextAware {

    private ApplicationContext applicationContext;

    private final Map<Class<? extends Request>,
            Class<? extends Handler<? extends Request, ? extends Response>>> handlerMap;

    public Mediator() {
        this.handlerMap = new HashMap<>();
        handlerMap.put(ReadUserListsQuery.class, ReadUserListsQueryHandler.class);
        handlerMap.put(ReadListItemsQuery.class, ReadListItemsQueryHandler.class);
        handlerMap.put(ReadUserWasteQuery.class, ReadUserWasteQueryHandler.class);
        handlerMap.put(ReadReportQuery.class, ReadReportQueryHandler.class);
        handlerMap.put(ReadAllUsersQuery.class, ReadAllUsersQueryHandler.class);
        handlerMap.put(AddListCommand.class, AddListCommandHandler.class);
        handlerMap.put(AddItemCommand.class, AddItemCommandHandler.class);
        handlerMap.put(DeleteItemCommand.class, DeleteItemCommandHandler.class);
        handlerMap.put(DeleteListCommand.class, DeleteListCommandHandler.class);
        handlerMap.put(LoginUserCommand.class, LoginUserCommandHandler.class);
        handlerMap.put(UpdateUserCommand.class, UpdateUserCommandHandler.class);
    }

    public <T extends Request, R extends Response> Handler<T, R> handle(T request) {
        Class<? extends Handler<? extends Request, ? extends Response>> handlerType = handlerMap.get(request.getClass());
        return (Handler<T, R>) applicationContext.getBean(handlerType);
    }

    @Override
    public void setApplicationContext(@NotNull ApplicationContext applicationContext) throws BeansException {
        this.applicationContext = applicationContext;
    }
}

File: WasteLessServer/src/main/java/server/mediator/Mediator.kt
package server.mediator

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.AnnotationConfigApplicationContext
import server.config.SpringConfig
import server.mediator.command.AddListCommand
import server.mediator.handler.AddListCommandHandler
import server.mediator.handler.Handler
import server.mediator.handler.ReadListItemsQueryHandler
import server.mediator.handler.ReadUserListsQueryHandler
import server.mediator.query.ReadListItemsQuery
import server.mediator.query.ReadUserListsQuery
import java.lang.reflect.Type
import kotlin.reflect.jvm.javaType
import kotlin.reflect.typeOf


@ExperimentalStdlibApi
class Mediator{

    @Autowired
    lateinit var handlerMap: HashMap<Type, Type>

    init {
        handlerMap[typeOf<AddListCommand>().javaType] = typeOf<AddListCommandHandler>().javaType
        handlerMap[typeOf<ReadUserListsQuery>().javaType] = typeOf<ReadUserListsQueryHandler>().javaType
        handlerMap[typeOf<ReadListItemsQuery>().javaType] = typeOf<ReadListItemsQueryHandler>().javaType
    }

    inline fun <reified TRequest: Request, reified TResponse> handle(request: TRequest): TResponse{
        val handlerType = handlerMap[typeOf<TRequest>().javaType]
        val context: ApplicationContext = AnnotationConfigApplicationContext(SpringConfig::class.java)
        val handler = context.getBean(handlerType!!::class.java) as Handler<TRequest, TResponse>

        return handler.handle(request)
    }

}